File: ios/pillarbox/Sources/pillarbox/PillarboxPlugin.swift
import Flutter
import UIKit

public class PillarboxPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        // TODO: Find flutter view controller correcly for add to app case
        let flutterViewController = UIApplication.shared.delegate?.window??.rootViewController as! FlutterViewController
        let factory = PillarboxNativeViewFactory(messenger: registrar.messenger(), flutterViewController: flutterViewController)
        registrar.register(factory, withId: "pillarbox-view")

        let channel = FlutterMethodChannel(name: "pillarbox", binaryMessenger: registrar.messenger())
        let instance = PillarboxPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)

        pluginregistrar = registrar
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "initialize":
            var arguments = call.arguments as! [String: Any]
            var dataSource = arguments["dataSource"] as! String
            var identifier = arguments["identifier"] as! Int
            controllers[identifier] = PillarboxController(identifier: identifier, uri: dataSource, messenger: pluginregistrar!.messenger())
            result("\(controllers.count) pillarbox controllers")
        case "dispose":
            var arguments = call.arguments as! [String: Any]
            var identifier = arguments["identifier"] as! Int
            controllers[identifier]?.player?.pause()
            controllers[identifier]?.player = nil
            controllers[identifier] = nil
            result("\(controllers.count) pillarbox controllers")
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

var controllers: [Int: PillarboxController] = [:]
private var pluginregistrar: FlutterPluginRegistrar!

File: android/src/main/kotlin/com/example/pillarbox/PillarboxPlugin.kt
package com.example.pillarbox

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** PillarboxPlugin */
class PillarboxPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        val factory = PillarboxNativeViewFactory(flutterPluginBinding.binaryMessenger)
        flutterPluginBinding.platformViewRegistry.registerViewFactory("pillarbox-view", factory)

        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "pillarbox")
        channel.setMethodCallHandler(this)

        binding = flutterPluginBinding
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "initialize") {
            val arguments = call.arguments as Map<*, *>
            val identifier = arguments["identifier"] as Int
            val uri = arguments["dataSource"] as String
            controllers[identifier] = PillarboxController(binding.applicationContext, identifier, uri, binding.binaryMessenger)
            result.success("${controllers.count()} pillarbox controllers")
        } else if (call.method == "dispose") {
            val arguments = call.arguments as Map<*, *>
            val identifier = arguments["identifier"] as Int
            controllers[identifier]?.player?.pause()
            controllers.remove(identifier)
            result.success("${controllers.count()} pillarbox controllers")
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

var controllers: MutableMap<Int, PillarboxController> = mutableMapOf()
lateinit var binding: FlutterPluginBinding